### PR TITLE
Set up ThreadedObserver and Buffers Structure in Robot Diagnostics

### LINKED ISSUE #1229: Setup ThreadedObserver and Buffers Structure in Robot Diagnostics
### Description of the task
Setup ThreadedObserver and Buffers structure in Robot Diagnostics. Currently Robot Diagnostics has stubs for communicating with the robot and the feedback area is undeveloped. Before we can handle communication, we need to first set up a structure for communicating data between the main thread and the gui thread. In addition, we need to make RobotDiagnostics an `Observer` of `SensorMsg` so it can get its data directly from `Backend` 

### Acceptance criteria

- [ ] Setup ThreadedObserver and Buffers structure in Robot Diagnostics
- [ ] Add more structure

### Blocked By
#1227 #1228

### PR DESCRIPTION
**Pull request by jonathanlew**
### Description
* Made RobotDiagnostics multithreaded in preparation for integrating Backend into main
* Setup ThreadedObserver and Buffers Structure in Robot Diagnostics
* Cleaned up callbacks and did refactoring

### Testing Done
Visual inspection/manual testing with print outs

### Resolved Issues
resolves #1229

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/gui/robot_diagnostics/widgets/robot_diagnostics.h [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+ * This class uses ThreadSafeBuffers to send/receive new data, and updates the GUI with
+ * the new data at a defined rate using a timer.
+ */
+class RobotDiagnostics : public QMainWindow
+{
+    Q_OBJECT
+
+   public:
+    /**
+     * Creates a new Visualizer MainWindow
```
Comment (MathewMacDougall) [inline src/software/gui/robot_diagnostics/widgets/robot_diagnostics.h]:
```suggestion
     * Creates a new RobotDiagnostics MainWindow
```

File: src/software/gui/robot_diagnostics/widgets/robot_diagnostics.h [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        std::shared_ptr<ThreadSafeBuffer<std::unique_ptr<Primitive>>> primitive_buffer,
+        QWidget* parent = nullptr);
+
+   private:
+    /**
+     * Push new primitive to buffer using new values received
+     *
+     * @param new values
+     */
+    void receiveChickerState(double chicker_power, ChargeMode charge_mode,
```
Comment (MathewMacDougall) [inline src/software/gui/robot_diagnostics/widgets/robot_diagnostics.h]: nit: These really feel more like callback functions to me. `receive` in the name makes me think we're getting data from an external source (sort of like the ThreadedObservers)

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
OVerall looks great, not much to say about the existing code.

I do want to quickly talk about how primitives will be sent out in the future. 

(1) Currently it looks like the system we're going for is to store a `current_primitive` of some kind and update parts of its value with the `receive...` functions as the user changes the UI. The primitive will be incrementally updated and sent out.

Because there may be several different primitives we want to send (Kick, Chip, DIrect wheels, direct velocity), it seems like we might need to store a "current primitive" for each type?

(2) The other alternative I can think of is converting the UI state into a primitive all at once. Eg. have any callback trigger "createAndSendPrimitive" that will create a primitive from the UI state all at once.

Pros of (1)
* granular callbacks useful for having certain UI changes trigger other changes.

Pros of (2)
* Don't have to store state for several primitives and have conditional logic for updating them (rather have conditional logic for creating them, _maybe_ simpler)

I think storing state for several types of primitives isn't an ideal solution so want to think ahead and brainstorm what this might look like ahead of time before we're too locked in to a callback/communication structure.

**Comment by jonathanlew** [comment]
> OVerall looks great, not much to say about the existing code.
> 
> I do want to quickly talk about how primitives will be sent out in the future.
> 
> (1) Currently it looks like the system we're going for is to store a `current_primitive` of some kind and update parts of its value with the `receive...` functions as the user changes the UI. The primitive will be incrementally updated and sent out.
> 
> Because there may be several different primitives we want to send (Kick, Chip, DIrect wheels, direct velocity), it seems like we might need to store a "current primitive" for each type?
> 
> (2) The other alternative I can think of is converting the UI state into a primitive all at once. Eg. have any callback trigger "createAndSendPrimitive" that will create a primitive from the UI state all at once.
> 
> Pros of (1)
> 
>     * granular callbacks useful for having certain UI changes trigger other changes.
> 
> 
> Pros of (2)
> 
>     * Don't have to store state for several primitives and have conditional logic for updating them (rather have conditional logic for creating them, _maybe_ simpler)
> 
> 
> I think storing state for several types of primitives isn't an ideal solution so want to think ahead and brainstorm what this might look like ahead of time before we're too locked in to a callback/communication structure.

Good point. I think we can go with (2). I think (1) would imply that we have simultaneous testing, i.e. we can have dribbling layers on top of driving, which I don't think we need to support.
Each callback should contain enough information to map a single primitive, e.g. dribbleprimitive, driveprimitive, etc., so any state can be static within the function, and then it can add that primitive to the buffer.

File: src/software/gui/robot_diagnostics/widgets/robot_diagnostics.h (hunk 2 above)
Comment (jonathanlew) [reply src/software/gui/robot_diagnostics/widgets/robot_diagnostics.h]: changed to on__changed

File: src/software/gui/robot_diagnostics/widgets/robot_diagnostics.h (hunk 1 above)
Comment (jonathanlew) [reply src/software/gui/robot_diagnostics/widgets/robot_diagnostics.h]: changed

**Comment by MathewMacDougall** [comment]
> Good point. I think we can go with (2). I think (1) would imply that we have simultaneous testing, i.e. we can have dribbling layers on top of driving, which I don't think we need to support.

I don't totally get what you mean by layers. You mean like "layering" functionality like dribbling + moving that isn't natively part of a primitive?

> Each callback should contain enough information to map a single primitive, e.g. dribbleprimitive, driveprimitive, etc., so any state can be static within the function, and then it can add that primitive to the buffer.

Just to be clear, it doesn't seem like that's what we have atm. We have more granular callbacks for each "subcomponent" like the dribbler, chicker, etc.
The callback triggered by changing `kick` to `chip` for example wouldn't contain all the info for a `Direct...` primitive since technically driving info needs to be included as well.

I think I was suggesting something more like each widget / component exposing a getter function for its current data eg. `getChargeState`, `getChickMode` etc. Then on a callback trigger you can construct a primitive by calling all these getter functions. The thing is that I like the callback system returning the "new" values for the component, but it would be patially redundant if we did a "getter" system like this because you could get the data from 2 places. That's why solution (2) is technically not as flexible if you go the pure getter route, because you can't have other things use the callbacks in as useful of a way.

**Review by garethellis0 — APPROVED** [review]
Were there any GUI changes in this? Just lemme know either way so I can build locally and look at it if so.

Otherwise LGTM :+1: :tada:

**Comment by jonathanlew** [comment]
> Just to be clear, it doesn't seem like that's what we have atm. We have more granular callbacks for each "subcomponent" like the dribbler, chicker, etc.
> The callback triggered by changing `kick` to `chip` for example wouldn't contain all the info for a `Direct...` primitive since technically driving info needs to be included as well.

What I mean by layering is why should the callback for kick/chip send a primitive for driving? It _could_ just send a primitive for kicking or chipping. However, if you set some drive settings and then press kick, it would suddenly stop driving (the drive widget would have to reset to give a visual indication, but that shouldn't be too hard). Maybe that's not behaviour we want to see. If we want driving and kicking at the same time, then we need to have separate getters probably.

**Comment by jonathanlew** [comment]
> Were there any GUI changes in this? Just lemme know either way so I can build locally and look at it if so.
> 
> Otherwise LGTM +1 tada

No GUI changes :smile:

**Comment by MathewMacDougall** [comment]
> What I mean by layering is why should the callback for kick/chip send a primitive for driving? It _could_ just send a primitive for kicking or chipping. However, if you set some drive settings and then press kick, it would suddenly stop driving (the drive widget would have to reset to give a visual indication, but that shouldn't be too hard). Maybe that's not behaviour we want to see. If we want driving and kicking at the same time, then we need to have separate getters probably.

Yea I was operating under the assumption we were using our current primitives as they are, which contain driving and kicking info together. I don't really think creating separate primitives for just kicking and chipping is a great idea (I don't see a ton of value in making specific primitives only used for testing when we already have the same functionality). You could at least get the same behavior by "default constructing" a `DirectVelocity` primitive (so it doesn't move) and just setting the kick/chip info.

I think driving and kicking at the same time is somewhat reasonable as a test case we'd want to support. For example, maybe elec would want to check how much current is being drawn in that moment to see if the system can handle it / it's safe. Or if kicking creates a bunch of electrical noise.

Thoughts?

**Comment by jonathanlew** [comment]
> I think driving and kicking at the same time is somewhat reasonable as a test case we'd want to support. For example, maybe elec would want to check how much current is being drawn in that moment to see if the system can handle it / it's safe. Or if kicking creates a bunch of electrical noise.
> 
> Thoughts?

That's a fair point. Then we should support dribbling at the same time too.

**Comment by MathewMacDougall** [comment]
> That's a fair point. Then we should support dribbling at the same time too.

:+1: I'm fine making changes to support this in a new ticket / PR so we can plan it out. This PR still adds lots of value in the wrappers and stuff that we might as well get in now.

**Review by MathewMacDougall — APPROVED** [review]

**Review by LiCody — APPROVED** [review]
LGTM 🎉

**Pull request merged**